### PR TITLE
Add script to automatically update changelog

### PR DESCRIPTION
**Pull request by charlesbvll**
### Description

Currently we have to manually update the changelog with all new changes since the previous release. This is very tedious.

### Related issues/PRs

N/A

### Explanation

This new script allows us to add changelog entries based on all new PRs' descriptions.
By default, it will search for a `Changelog entry` section in the description of the PR, if it is empty, it will just use the title of the PR for the changelog entry, without any description. If it is non existent, it will categorize the PR as "General improvement" and add it to the changelog accordingly. If it exists, it will use it to add a description to the change.

Note that a system of tokens is also in place:
- `<general>` is for classifying a PR as a general improvement.
- `<skip>` is to not add the PR to the changelog
- `<baselines>` is to add a general baselines change to the PR
- `<examples>` is to add a general examples change to the PR
- `<sdk>` is to add a general sdk change to the PR
- `<simulations>` is to add a general simulations change to the PR

Once those tokens are detected, if the corresponding entry (like `- **General improvements**`) doesn't already exist, it gets automatically created, and otherwise, PRs' URLs are added at the end of the list inside the parenthesis.

### Checklist

- [x] Implement proposed change
- [x] Update changelog entry below
- [x] Make CI checks pass
- [x] Ping maintainers on [Slack](https://flower.dev/join-slack/) (channel `#contributions`)

### Changelog entry

<skip>

### Any other comments?

In the new PR template, <general> is the default flag.

### CODE REVIEW

File: src/py/flwr_tool/update_changelog.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+    }
+    return headers.get(category, "")
+
+
+def _update_entry(
+    content, category_title, pr_info, unreleased_index, next_header_index
+):
+    """Update a specific section in the changelog content."""
+    section_index = content.find(category_title, unreleased_index, next_header_index)
+    if section_index != -1:
```
Comment (Robert-Steiner) [inline src/py/flwr_tool/update_changelog.py:162]: just want to mention a small tick i have discovered recently. the two lines can be combined in one by using the walrus operator `:=` that has been introduced in python 3.8.

(i haven't tested this line)
```suggestion
    if (section_index := content.find(category_title, unreleased_index, next_header_index)) != -1:
```
but i have no strong preference using it here so we can also leave the code as is.

**Review by Robert-Steiner — COMMENTED** [review]
Great work! I really like the script 👍

File: src/py/flwr_tool/update_changelog.py (hunk 1 above)
Comment (charlesbvll) [reply src/py/flwr_tool/update_changelog.py:162]: Oh nice! I'll test it out, thanks!

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -41,6 +41,10 @@ Example: The variable `rnd` was renamed to `server_round` to improve readability
 - [ ] Make CI checks pass
 - [ ] Ping maintainers on [Slack](https://flower.dev/join-slack/) (channel `#contributions`)
 
+## Changelog entry
+
+<general>
```
[the remark this answers is not in the document]
Comment (charlesbvll) [reply .github/PULL_REQUEST_TEMPLATE.md:46]: That's a good point, I will add a comment explaining everything above the section

**Comment by tanertopal** [comment]
@charlesbvll This is cool! I believe it would be helpful to write a small guide in the docs "CONTRIBUTOR HOW-TO GUIDES" section on how this works, with a few examples to send to new contributors to explain it.

@danieljanes @Robert-Steiner thoughts?

**Comment by charlesbvll** [comment]
> @charlesbvll This is cool! I believe it would be helpful to write a small guide in the docs "CONTRIBUTOR HOW-TO GUIDES" section on how this works, with a few examples to send to new contributors to explain it.
> 
> @danieljanes @Robert-Steiner thoughts?

I just added a quick mention at the moment, where we talk about the description of the PR, but maybe I can add a new section at the end of the doc, like an appendix, giving more details and some examples.

**Comment by charlesbvll** [comment]
> > @charlesbvll This is cool! I believe it would be helpful to write a small guide in the docs "CONTRIBUTOR HOW-TO GUIDES" section on how this works, with a few examples to send to new contributors to explain it.
> > @danieljanes @Robert-Steiner thoughts?
> 
> I just added a quick mention at the moment, where we talk about the description of the PR, but maybe I can add a new section at the end of the doc, like an appendix, giving more details and some examples.

@tanertopal I added an appendix, I think the PR is ready for review

**Review by Robert-Steiner — COMMENTED** [review]
From my side it looks good!

**Review by tanertopal — APPROVED** [review]

**Pull request merged**